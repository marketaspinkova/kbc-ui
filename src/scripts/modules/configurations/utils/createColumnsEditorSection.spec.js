import assert from 'assert';
import React, {PropTypes} from 'react';
import createColumnsEditorSection from './createColumnsEditorSection';
import {fromJS} from 'immutable';

const ColumnMapping = ({column, disabled, onChange, showAdvanced, context}) => (
  <span disabled={disabled}>
    <div>{column.columnId}</div>
    <div>{column.type}</div>
    <div>{context.tableId}</div>
    <div>{context.columnsCount}</div>
    <span onClick={() => onChange({...column, type: 'ignore'})}> ignore column</span>
    {showAdvanced && <span> show advanced</span>}
  </span>
);

ColumnMapping.propTypes = {
  onChange: PropTypes.func.isRequired,
  disabled: PropTypes.bool.isRequired,
  column: PropTypes.object.isRequired,
  context: PropTypes.object.isRequired,
  showAdvanced: PropTypes.bool
};


const configuration = fromJS({
  parameters: {
    storage: {
      input: {
        tables: [{
          source: 'in.some.table',
          columns: ['age', 'person', 'deletedColumn']
        }]
      }
    },
    tableId: 'in.some.table',
    columns: [
      {
        columnId: 'age',
        type: 'number'
      },
      {
        columnId: 'person',
        type: 'varchar'
      },
      {
        columnId: 'deletedColumn',
        type: 'varchar'
      }
    ]
  }
});

const sectionContext = fromJS({
  table: {
    id: 'in.some.table',
    columns: ['age', 'person', 'otherColumn']
  }
});

const editorSectionDefinition = {
  matchColumnKey: 'columnId',
  onLoadColumns: (config) => config.getIn(['parameters', 'columns']),
  onSaveColumns: (tableId, columnsList) => fromJS({
    parameters: {
      tableId: tableId,
      columns: columnsList
    }
  }),
  isColumnIgnored: column => column.get('type') === 'ignore',
  initColumnFn: columnName => fromJS({columnId: columnName, type: 'ignore'}),
  isColumnValidFn: column => ['varchar', 'number'].includes(column.type),
  prepareColumnContext: (context, allColumns) => fromJS({
    tableId: context.getIn(['table', 'id']),
    columnsCount: allColumns.count()
  }),
  getInitialShowAdvanced: () => false,
  columnsMappings: [
    {
      title: 'Column Setup',
      render: ColumnMapping
    }
  ]
};

describe('columns editor section', () => {
  it('test onLoad and onSave section functions', () => {
    const editorSection = createColumnsEditorSection(editorSectionDefinition).toJS();
    const localState = editorSection.onLoad(configuration, sectionContext);
    assert.deepEqual(
      localState.toJS(),
      {
        columns: [
          {
            columnId: 'age',
            type: 'number'
          },
          {
            columnId: 'person',
            type: 'varchar'
          },
          {
            columnId: 'otherColumn',
            type: 'ignore'
          },
          {
            columnId: 'deletedColumn',
            type: 'varchar'
          }
        ],
        tableId: 'in.some.table',
        columnsMappings: editorSectionDefinition.columnsMappings,
        context: {
          tableId: 'in.some.table',
          columnsCount: 4
        },
        matchColumnKey: 'columnId',
        getInitialShowAdvanced: editorSectionDefinition.getInitialShowAdvanced,
        isColumnValidFn: editorSectionDefinition.isColumnValidFn
      }
    );

    const configToSave = editorSection.onSave(localState);
    assert.deepEqual(
      configToSave.toJS(),
      {
        storage: {
          input: {
            tables: [
              {
                source: 'in.some.table',
                columns: [
                  'age',
                  'person',
                  'deletedColumn'
                ]
              }
            ]
          }
        },
        parameters: {
          tableId: 'in.some.table',
          columns: [
            {
              columnId: 'age',
              type: 'number'
            },
            {
              columnId: 'person',
              type: 'varchar'
            },
            {
              columnId: 'deletedColumn',
              type: 'varchar'
            }
          ]
        }
      }
    );
  });
  it('test test render', () => {
    const editorSection = createColumnsEditorSection(editorSectionDefinition).toJS();
    const localState = editorSection.onLoad(configuration, sectionContext).toJS();
    const EditorComponent = editorSection.render;
    renderSnapshot(
      <EditorComponent
        value={localState}
        onChange={() => null}
        disabled={false}/>
    );
  });
});
