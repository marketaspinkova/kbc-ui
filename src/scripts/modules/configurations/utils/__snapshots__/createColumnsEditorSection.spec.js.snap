// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`columns editor section test nonexisting table render 1`] = `
<div>
  <div
    class="alert alert-warning"
  >
    <h3>
      Table Missing
    </h3>
    <div
      class="help-block"
    >
      <span>
        Table 
        <code>
          in.some.table
        </code>
         used in this configuration is missing in Storage. Running this configuration will fail.
      </span>
    </div>
  </div>
  <div>
    <h3>
      Columns
    </h3>
    <div
      class="storage-table-columns-editor-wrapper"
    >
      <table
        class="storage-table-columns-editor table table-striped"
      >
        <thead>
          <tr>
            <th
              class="col-md-2"
            >
              Column Name
            </th>
            <th>
              Column Setup
            </th>
            <th
              class="col-md-1"
            >
              Preview
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
          >
            <td
              class="column-name"
            >
              age
            </td>
            <td>
              <span>
                <div>
                  age
                </div>
                <div>
                  number
                </div>
                <div />
                <div>
                  3
                </div>
                <span>
                   ignore column
                </span>
              </span>
            </td>
            <td>
              <button
                class="btn btn-link"
                type="button"
              >
                <i
                  class="fa fa-eye"
                />
              </button>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="column-name"
            >
              person
            </td>
            <td>
              <span>
                <div>
                  person
                </div>
                <div>
                  varchar
                </div>
                <div />
                <div>
                  3
                </div>
                <span>
                   ignore column
                </span>
              </span>
            </td>
            <td>
              <button
                class="btn btn-link"
                type="button"
              >
                <i
                  class="fa fa-eye"
                />
              </button>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="column-name"
            >
              deletedColumn
            </td>
            <td>
              <span>
                <div>
                  deletedColumn
                </div>
                <div>
                  varchar
                </div>
                <div />
                <div>
                  3
                </div>
                <span>
                   ignore column
                </span>
              </span>
            </td>
            <td>
              <button
                class="btn btn-link"
                type="button"
              >
                <i
                  class="fa fa-eye"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`columns editor section test render 1`] = `
<div>
  <div>
    <h3>
      Columns
    </h3>
    <div
      class="storage-table-columns-editor-wrapper"
    >
      <table
        class="storage-table-columns-editor table table-striped"
      >
        <thead>
          <tr>
            <th
              class="col-md-2"
            >
              Column Name
            </th>
            <th>
              Column Setup
            </th>
            <th
              class="col-md-1"
            >
              Preview
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
          >
            <td
              class="column-name"
            >
              age
            </td>
            <td>
              <span>
                <div>
                  age
                </div>
                <div>
                  number
                </div>
                <div>
                  in.some.table
                </div>
                <div>
                  4
                </div>
                <span>
                   ignore column
                </span>
              </span>
            </td>
            <td>
              <button
                class="btn btn-link"
                type="button"
              >
                <i
                  class="fa fa-eye"
                />
              </button>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="column-name"
            >
              person
            </td>
            <td>
              <span>
                <div>
                  person
                </div>
                <div>
                  varchar
                </div>
                <div>
                  in.some.table
                </div>
                <div>
                  4
                </div>
                <span>
                   ignore column
                </span>
              </span>
            </td>
            <td>
              <button
                class="btn btn-link"
                type="button"
              >
                <i
                  class="fa fa-eye"
                />
              </button>
            </td>
          </tr>
          <tr
            class="danger"
          >
            <td
              class="column-name"
            >
              otherColumn
            </td>
            <td>
              <span>
                <div>
                  otherColumn
                </div>
                <div>
                  ignore
                </div>
                <div>
                  in.some.table
                </div>
                <div>
                  4
                </div>
                <span>
                   ignore column
                </span>
              </span>
            </td>
            <td>
              <button
                class="btn btn-link"
                type="button"
              >
                <i
                  class="fa fa-eye"
                />
              </button>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="column-name"
            >
              deletedColumn
            </td>
            <td>
              <span>
                <div>
                  deletedColumn
                </div>
                <div>
                  varchar
                </div>
                <div>
                  in.some.table
                </div>
                <div>
                  4
                </div>
                <span>
                   ignore column
                </span>
              </span>
            </td>
            <td>
              <button
                class="btn btn-link"
                type="button"
              >
                <i
                  class="fa fa-eye"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
